File: frontend/src/api/patrimonio.ts
import api from './client'

export type TipoBem = 'movel' | 'imovel'
export type ObjetivoBem = 'venda' | 'aluguel' | 'uso_proprio' | 'uso_herdeiro' | 'nao_fazer_nada'
export type StatusBem = 'em_validacao' | 'validado' | 'incerto'
export type OrigemTitulo = 'matricula_rgi' | 'escritura_publica' | 'contrato_particular' | 'outro'
export type TipoDocumentoElo =
  | 'contrato_compra_venda'
  | 'escritura_publica'
  | 'cessao_direitos'
  | 'matricula'
  | 'formal_partilha'
  | 'outro'

export interface Anexo {
  id: string
  filename: string
  drive_link?: string | null
  mime?: string | null
  created_at: string
}

export interface CadeiaElo {
  id: string
  ordem: number
  tipo_documento: TipoDocumentoElo
  de_quem?: string | null
  para_quem?: string | null
  data?: string | null
  descricao?: string | null
  arquivo_nome?: string | null
  drive_link?: string | null
  created_at: string
}

export interface Socio {
  id: string
  ordem: number
  nome: string
  cpf?: string | null
  percentual?: number | null
  integralizar: boolean
  created_at: string
}

export interface Bem {
  id: string
  cliente_id: string
  tipo_bem: TipoBem
  nome: string
  descricao?: string | null
  valor_compra?: number | null
  valor_mercado?: number | null
  valor_ir?: number | null
  data_compra?: string | null
  objetivo?: ObjetivoBem | null
  descricao_matricula?: string | null
  numero_matricula?: string | null
  cartorio?: string | null
  origem_titulo?: OrigemTitulo | null
  escritura_numero?: string | null
  escritura_livro?: string | null
  escritura_folha?: string | null
  ordem?: number | null
  status: StatusBem
  integralizar_holding: boolean
  proprietario_real?: string | null
  proprietario_matricula?: string | null
  tem_gravame: boolean
  gravame_descricao?: string | null
  observacoes?: string | null
  empresa_nome?: string | null
  empresa_cnpj?: string | null
  capital_social?: number | null
  valor_balanco?: number | null
  data_balanco?: string | null
  participacao_cliente_pct?: number | null
  anexos: Anexo[]
  cadeia: CadeiaElo[]
  socios: Socio[]
  created_at: string
  updated_at: string
}

export type BemCreate = { cliente_id: string } & Partial<Omit<Bem, 'id' | 'cliente_id' | 'anexos' | 'cadeia' | 'created_at' | 'updated_at'>> & { nome: string }
export type BemUpdate = Partial<Omit<Bem, 'id' | 'cliente_id' | 'anexos' | 'cadeia' | 'created_at' | 'updated_at'>>

export type CadeiaEloCreate = Partial<Omit<CadeiaElo, 'id' | 'arquivo_nome' | 'drive_link' | 'created_at'>>
export type CadeiaEloUpdate = CadeiaEloCreate

export type SocioCreate = Partial<Omit<Socio, 'id' | 'created_at'>> & { nome: string }
export type SocioUpdate = Partial<Omit<Socio, 'id' | 'created_at'>>

export interface EscrituraCampo {
  valor: string | number | null
  trecho: string
  referencia?: string
}
export interface GravameItem {
  tipo: string
  descricao: string
  vencida: boolean | null
  trecho: string
  referencia?: string
}
export interface EscrituraExtraida {
  descricao_imovel: EscrituraCampo
  proprietario_atual: EscrituraCampo & { data_aquisicao?: string | null }
  valor_compra: EscrituraCampo & { moeda?: string }
  proprietarios_anteriores: EscrituraCampo
  data_transacao: EscrituraCampo
  numero_matricula: EscrituraCampo
  cartorio: EscrituraCampo
  gravames: { existe: boolean; itens: GravameItem[] }
}

export const patrimonioApi = {
  extrairEscritura: (file: File) => {
    const form = new FormData()
    form.append('file', file)
    return api.post<EscrituraExtraida>('/patrimonio/extrair-escritura', form, {
      timeout: 200000,
    }).then((r) => r.data)
  },

  listar: (clienteId: string) =>
    api.get<Bem[]>('/patrimonio/', { params: { cliente_id: clienteId } }).then((r) => r.data),

  criar: (data: BemCreate) => api.post<Bem>('/patrimonio/', data).then((r) => r.data),

  atualizar: (id: string, data: BemUpdate) =>
    api.patch<Bem>(`/patrimonio/${id}`, data).then((r) => r.data),

  deletar: (id: string) => api.delete(`/patrimonio/${id}`),

  reordenar: (clienteId: string, ordem: string[]) =>
    api.post(`/patrimonio/reordenar`, { ordem }, { params: { cliente_id: clienteId } }).then((r) => r.data),

  uploadAnexo: (bemId: string, file: File) => {
    const form = new FormData()
    form.append('file', file)
    return api.post<Anexo>(`/patrimonio/${bemId}/anexos`, form).then((r) => r.data)
  },

  deletarAnexo: (bemId: string, anexoId: string) =>
    api.delete(`/patrimonio/${bemId}/anexos/${anexoId}`),

  criarElo: (bemId: string, data: CadeiaEloCreate) =>
    api.post<CadeiaElo>(`/patrimonio/${bemId}/cadeia`, data).then((r) => r.data),

  atualizarElo: (bemId: string, eloId: string, data: CadeiaEloUpdate) =>
    api.patch<CadeiaElo>(`/patrimonio/${bemId}/cadeia/${eloId}`, data).then((r) => r.data),

  deletarElo: (bemId: string, eloId: string) =>
    api.delete(`/patrimonio/${bemId}/cadeia/${eloId}`),

  uploadAnexoElo: (bemId: string, eloId: string, file: File) => {
    const form = new FormData()
    form.append('file', file)
    return api.post<CadeiaElo>(`/patrimonio/${bemId}/cadeia/${eloId}/anexo`, form).then((r) => r.data)
  },

  criarSocio: (bemId: string, data: SocioCreate) =>
    api.post<Socio>(`/patrimonio/${bemId}/socios`, data).then((r) => r.data),

  atualizarSocio: (bemId: string, socioId: string, data: SocioUpdate) =>
    api.patch<Socio>(`/patrimonio/${bemId}/socios/${socioId}`, data).then((r) => r.data),

  deletarSocio: (bemId: string, socioId: string) =>
    api.delete(`/patrimonio/${bemId}/socios/${socioId}`),

  exportXls: (clienteId: string) =>
    api.get(`/patrimonio/export/xls`, { params: { cliente_id: clienteId }, responseType: 'blob' })
      .then((r) => r.data as Blob),

  exportPdf: (clienteId: string) =>
    api.get(`/patrimonio/export/pdf`, { params: { cliente_id: clienteId }, responseType: 'blob' })
      .then((r) => r.data as Blob),
}

File: backend/app/schemas/patrimonio.py
import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel

TipoBem = Literal["movel", "imovel"]
Objetivo = Literal["venda", "aluguel", "uso_proprio", "uso_herdeiro", "nao_fazer_nada", "segurar"]
StatusBem = Literal["em_validacao", "validado", "incerto"]
OrigemTitulo = Literal["matricula_rgi", "escritura_publica", "contrato_particular", "outro"]
TipoDocumentoElo = Literal[
    "contrato_compra_venda",
    "escritura_publica",
    "cessao_direitos",
    "matricula",
    "formal_partilha",
    "outro",
]


# ── Anexos ──────────────────────────────────────────────────────────────────
class AnexoOut(BaseModel):
    id: uuid.UUID
    filename: str
    drive_link: str | None = None
    mime: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Cadeia sucessória ───────────────────────────────────────────────────────
class CadeiaEloBase(BaseModel):
    ordem: int = 0
    tipo_documento: TipoDocumentoElo = "outro"
    de_quem: str | None = None
    para_quem: str | None = None
    data: date | None = None
    descricao: str | None = None


class CadeiaEloCreate(CadeiaEloBase):
    pass


class CadeiaEloUpdate(BaseModel):
    ordem: int | None = None
    tipo_documento: TipoDocumentoElo | None = None
    de_quem: str | None = None
    para_quem: str | None = None
    data: date | None = None
    descricao: str | None = None


class CadeiaEloOut(CadeiaEloBase):
    id: uuid.UUID
    arquivo_nome: str | None = None
    drive_link: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Sócios (bem móvel = cota social) ─────────────────────────────────────────
class SocioBase(BaseModel):
    ordem: int = 0
    nome: str
    cpf: str | None = None
    percentual: float | None = None
    integralizar: bool = False


class SocioCreate(SocioBase):
    pass


class SocioUpdate(BaseModel):
    ordem: int | None = None
    nome: str | None = None
    cpf: str | None = None
    percentual: float | None = None
    integralizar: bool | None = None


class SocioOut(SocioBase):
    id: uuid.UUID
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Comentários ──────────────────────────────────────────────────────────────
class ComentarioCreate(BaseModel):
    titulo: str | None = None
    texto: str


class ComentarioOut(BaseModel):
    id: uuid.UUID
    titulo: str | None = None
    texto: str
    autor_nome: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Bem ─────────────────────────────────────────────────────────────────────
class BemBase(BaseModel):
    tipo_bem: TipoBem = "imovel"
    nome: str
    descricao: str | None = None
    valor_compra: float | None = None
    valor_mercado: float | None = None
    valor_ir: float | None = None
    data_compra: date | None = None
    objetivo: Objetivo | None = None
    descricao_matricula: str | None = None
    numero_matricula: str | None = None
    cartorio: str | None = None
    origem_titulo: OrigemTitulo | None = None
    escritura_numero: str | None = None
    escritura_livro: str | None = None
    escritura_folha: str | None = None
    status: StatusBem = "em_validacao"
    integralizar_holding: bool = False
    proprietario_real: str | None = None
    proprietario_matricula: str | None = None
    tem_gravame: bool = False
    gravame_descricao: str | None = None
    observacoes: str | None = None
    # Cota social / participação societária (bem móvel)
    empresa_nome: str | None = None
    empresa_cnpj: str | None = None
    capital_social: float | None = None
    valor_balanco: float | None = None
    data_balanco: date | None = None
    participacao_cliente_pct: float | None = None


class BemCreate(BemBase):
    cliente_id: uuid.UUID


class BemUpdate(BaseModel):
    tipo_bem: TipoBem | None = None
    nome: str | None = None
    descricao: str | None = None
    valor_compra: float | None = None
    valor_mercado: float | None = None
    valor_ir: float | None = None
    data_compra: date | None = None
    objetivo: Objetivo | None = None
    descricao_matricula: str | None = None
    numero_matricula: str | None = None
    cartorio: str | None = None
    origem_titulo: OrigemTitulo | None = None
    escritura_numero: str | None = None
    escritura_livro: str | None = None
    escritura_folha: str | None = None
    status: StatusBem | None = None
    integralizar_holding: bool | None = None
    proprietario_real: str | None = None
    proprietario_matricula: str | None = None
    tem_gravame: bool | None = None
    gravame_descricao: str | None = None
    observacoes: str | None = None
    empresa_nome: str | None = None
    empresa_cnpj: str | None = None
    capital_social: float | None = None
    valor_balanco: float | None = None
    data_balanco: date | None = None
    participacao_cliente_pct: float | None = None


class BemOut(BemBase):
    id: uuid.UUID
    cliente_id: uuid.UUID
    ordem: int | None = None
    anexos: list[AnexoOut] = []
    cadeia: list[CadeiaEloOut] = []
    socios: list[SocioOut] = []
    comentarios: list[ComentarioOut] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
